### PR TITLE
type raw array and add precise inference rule for objectget and newob…

### PR DESCRIPTION
**Pull request by jackbackrack**
Typed raw array so that in optimized mode the element types are propagated properly.  Added a couple type inference rules so that the element types and object types are precisely inferred.

### CODE REVIEW

**Comment by CuppoJava** [comment]
Merged, the RawArray parametric types, but still need to review the changes to the type inference.

File: compiler/stz-el.stanza [hunk 1]
```diff
@@ -1866,14 +1866,14 @@ defn annotate-body-instructions (e:EBody, state0:Table<Int,EType>, vt:VarTable)
...
       (i:EObject) : trusted(x(i))
+      (i:EObjectGet) : trusted(x(i))
       (i:EArray) : trusted(x(i))
       (i:EStruct) : trusted(x(i))
       (i:EPtr) : trusted(x(i))
       (i:ELoad) : untrusted(x(i))
       (i:EInterpret) : untrusted(x(i))
       (i:EConv) : untrusted(x(i))
       (i:EPrim) : trusted(x(i))
-      (i:ENewObject) : trusted(x(i))
```
Comment (CuppoJava) [inline compiler/stz-el.stanza]: What was the reasoning for moving this out of this section and into the next?

File: compiler/stz-el.stanza [hunk 2]
```diff
@@ -1866,14 +1866,14 @@ defn annotate-body-instructions (e:EBody, state0:Table<Int,EType>, vt:VarTable)
       (i:EVoidTuple) : trusted(x(i))
       (i:ETupleGet) : untrusted(x(i))
       (i:EObject) : trusted(x(i))
+      (i:EObjectGet) : trusted(x(i))
```
Comment (CuppoJava) [inline compiler/stz-el.stanza:1869]: Looks good.

File: compiler/stz-el.stanza (hunk 1 above)
Comment (jackbackrack) [reply compiler/stz-el.stanza]: i tested and found in fact it doesn't affect the results to leave it the way it was.  the important thing is to have rawarray be typed it appears.

File: compiler/stz-el.stanza [hunk 3]
```diff
@@ -89,6 +89,8 @@ defn lower (epackage:EPackage, optimize?:True|False) -> EPackage :
   run-pass("Lift Closures", lift-closures, "closurelifted", false)
   run-pass("Lift Type Objects", lift-type-objects, "typelifted", false)
 
+  spit("out.txt", cur-package)
```
Comment (CuppoJava) [inline compiler/stz-el.stanza]: Can you get rid of the debug output?

**Review by CuppoJava — CHANGES_REQUESTED** [review]

**Comment by jackbackrack** [comment]
done.

**Pull request merged**